### PR TITLE
FeatureFlags: manage feature flags outside of settings.Cfg

### PR DESCRIPTION
**Pull request by ryantxu**
#### Goals:
* single/standard way to define a feature flag
* clear way to see what flags exist and basic information about them
* add more constraints for some flags.  IE: require running in dev mode, or with enterprise codebase
* easily trace where feature toggles are used (ie IDE "show references")

#### This PR:
1. creates a list of flags in the backend with additional metadata (name, description, state, etc)
2. generates typescript code that matches the flags
3. moves all featureToggle functions out of `setting.go` and into a `FeatureToggle` class
4. Only sends the enabled flags to the frontend.  Updates the types so we use `flag?: boolean` rather than force a complete object 
5. Makes sure flags that require enterprise license are enabled in the license

#### FUTURE: #43941
1. watch `features.yaml` for changes
2. compile real expressions (currently only "true" works) based on environment and variables
3. show simple UI for all flag states

#### FUTURE future
1. standardize the flag id naming conventions.  require camel|snake case?  avoid dashes?
2. unify various cfg 'enabled' settings as feature flags?
3. Well defined feature flag lifecycle, etc

### CODE REVIEW

File: packages/grafana-data/src/types/config.ts
Comment (ryantxu) [inline packages/grafana-data/src/types/config.ts:45]: @mckn -- any thoughts on why/if this needs to bee in `@grafana/data`?  It seems like most of the config stuff really belongs in runtime

File: pkg/setting/setting_feature_toggles.go
Comment (bergquist) [inline pkg/setting/setting_feature_toggles.go:48]: I think this should be part of the state instead. Once the feature is mature enough to be tested in production we have to avoid breaking changes to the schema.

File: pkg/setting/setting_feature_toggles.go
Comment (bergquist) [inline pkg/setting/setting_feature_toggles.go:38]: Is this really needed? I'm worried that it makes things come complicated. Accidentally enabling/disabling feature is extremely annoying :)

File: pkg/setting/setting_feature_toggles.go
Comment (bergquist) [inline pkg/setting/setting_feature_toggles.go:49]: Reloading feature without restarting should be supported by both frontend and backend. FGAC is a good example of a feature that could enable at runtime.

**Review by bergquist — COMMENTED** [review]
:smile: 

Glad to see more thoughts on Toggles! 

During the hackaton we implemented (but didn't get far due to incidents/on-call) https://github.com/grafana/grafana/tree/experimentals/feature-toggle-by-headers which introduces a new framework rather then building on top on what we have. The main reason for that is to make the distinction between a feature being in development mode (tested locally, database migrations) and ready for production/staging more clear. The new "experiments" can MUST always be possible to reload at runtime. But this could also be expressed in a state which you introduce in this PR. 

Enabling toggles without restarting the process is *extremely* valuable when operating Grafana at scale and that's why I favor a new config file(s) so we can mount the same file for multiple instances. 

An interesting idea about having states per toggles would be to allow admins to enable features above a certain state from within Grafana itself.

File: pkg/setting/setting_feature_toggles.go
Comment (ryantxu) [reply pkg/setting/setting_feature_toggles.go:47]: yes -- though it is also a question of who the metadata is for.  IE unclear to me if I set "accesscontrol" feature flag if that has any affect in OSS or not

Again -- my motivation for this was to try to understand what flags we currently have a basic understanding what they do

File: pkg/setting/setting_feature_toggles.go
Comment (ryantxu) [reply pkg/setting/setting_feature_toggles.go:38]: AliasIds?  -- not necessary now now.  It was copied from the "registry" structure we use everywhere in the frontend.  This essentially allows us to fix/clean/rename things without breaking existing configs.

But agree we can leave it off

File: pkg/setting/setting_feature_toggles.go
Comment (ryantxu) [reply pkg/setting/setting_feature_toggles.go:49]: :+1:

File: pkg/setting/setting_feature_toggles.go
Comment (ryantxu) [reply pkg/setting/setting_feature_toggles.go:48]: ie: state=alpha-DB?

FYI: I actually started this effort when starting a feature that adds a SQL table, and was looking for a good way to shout that.  My assumption is that if it got merged to main it is reasonable if future "migration" actually drops all data first :grimacing:

File: pkg/services/featuremgmt/manager.go
Comment (ryantxu) [inline pkg/services/featuremgmt/manager.go:79]: @bergquist -- TODO... right now this only supports the "true" expression!

File: pkg/services/featuremgmt/manager.go
Comment (ryantxu) [inline pkg/services/featuremgmt/manager.go:74]: @DanCech -- this will make sure any features that require a license are checked with the license provider

File: packages/grafana-runtime/src/utils/licensing.ts
Comment (ryantxu) [inline packages/grafana-runtime/src/utils/licensing.ts:4]: @DanCech -- this now uses the same `featureToggle` bucket, but note that the license is checked on the backend before anythign can be enabled

File: pkg/models/licensing.go
Comment (ryantxu) [inline pkg/models/licensing.go:17]: @DanCech -- I removed the enumeration from the main interface.  Is the map really necessary?

File: pkg/setting/setting.go
Comment (ryantxu) [inline pkg/setting/setting.go:346]: The circular dependencies make this pretty hard to get rid of :grimacing: 

Most places that used the map reference can be replaced with a refernce to `featuremgmt.FeatureToggles`, but a few places can not.

In particular the `settings.Provider` can not use featuremgmt without a circular ref

File: pkg/models/licensing.go
Comment (ryantxu) [reply pkg/models/licensing.go:17]: @DanCech -- do licenses ever set things to *false*?

File: pkg/setting/setting_feature_toggles_test.go
Comment (ryantxu) [inline pkg/setting/setting_feature_toggles_test.go:17]: @bergquist -- this behavior is now defined by the initial expr in `registry.go`

File: pkg/server/wire.go
Comment (ryantxu) [inline pkg/server/wire.go:185]: Perhaps over complicating this....  

Manager = can control (and eventually change) that flags
Toggles = readonly settings

File: pkg/models/licensing.go
Comment (DanCech) [reply pkg/models/licensing.go:17]: removing it is fine I think but we'll need to verify in enterprise, the map was intended to just be a list of true values, no false values.

File: pkg/server/wire.go
Comment (DanCech) [reply pkg/server/wire.go:185]: It's a reasonable distinction, I had been considering a follow-up PR for licensing to do something similar as the majority of consumers only need a very small subset of the current interface.

File: pkg/services/featuremgmt/manager_test.go
Comment (FZambia) [inline pkg/services/featuremgmt/manager_test.go:12]: I prefer using `github.com/stretchr/testify/require` instead of `assert` since in most cases first error is what we aiming for when looking at test failure, the following errors in test are usually similar or caused by the first one etc - i.e. not really deterministic after first one.

File: pkg/services/featuremgmt/toggles_gen_test.go
Comment (FZambia) [inline pkg/services/featuremgmt/toggles_gen_test.go:105]:
```suggestion
// Is{{.CamleCase}}Enabled() checks for the flag: {{.Flag.Name}}{{.Ext}}
```

File: pkg/services/featuremgmt/toggles_gen_test.go
Comment (FZambia) [inline pkg/services/featuremgmt/toggles_gen_test.go:115]: Typo in word CamleCase

File: pkg/services/featuremgmt/registry.go
Comment (FZambia) [inline pkg/services/featuremgmt/registry.go:86]: Can we somehow force the style for all next feature toggles, for example use camelCase only 🤔 Maybe at least write the suggested name format in FeatureFlag.Name field?

File: pkg/services/featuremgmt/features.go
Comment (FZambia) [inline pkg/services/featuremgmt/features.go:21]: What FeatureStateStable means btw? Does stable state means that this should be enabled by default or moved to its own config section from feature toggles list? Or it can stay here forever? It seems to me all feature toggles should be removed eventually.

File: pkg/services/featuremgmt/manager.go
Comment (FZambia) [inline pkg/services/featuremgmt/manager.go:133]:
```suggestion
// Toggles returns FeatureToggles.
```

File: pkg/services/featuremgmt/manager.go
Comment (FZambia) [inline pkg/services/featuremgmt/manager.go:38]: Should the check that add.Description is not empty be here? Moreover, maybe Description should be `*string` to allow overrides like this?

File: pkg/services/featuremgmt/service.go
Comment (FZambia) [inline pkg/services/featuremgmt/service.go:72]: 🤔

File: pkg/services/featuremgmt/toggles_gen_test.go
Comment (marefr) [inline pkg/services/featuremgmt/toggles_gen_test.go:18]: Nit. A bit weird to do this in a test. Should probably use go generate.

File: packages/grafana-data/src/types/featureToggles.gen.ts
Comment (bergquist) [inline packages/grafana-data/src/types/featureToggles.gen.ts:15]:
```suggestion
```
?

File: pkg/services/featuremgmt/toggles_gen_test.go
Comment (ryantxu) [reply pkg/services/featuremgmt/toggles_gen_test.go:115]: :facepalm:

File: pkg/services/featuremgmt/toggles_gen_test.go
Comment (ryantxu) [reply pkg/services/featuremgmt/toggles_gen_test.go:18]: my go generate foo is not strong... I see how to get it to create functions from static files, but no idea (and can not find any examples) where it generates typescript and go based on a staticly defined struct.

I added it as a test because we want to fail if things ever get out of sync

Comment (marefr) [reply pkg/services/featuremgmt/toggles_gen_test.go:18]: Okay :smile: We can look into that later if needed. There's probably many examples out on the www. It's just a lite weird that you need to run tests to generate the code and you want to commit the changes. In case you forget that you have to wait for CI to tell you about it

File: pkg/services/featuremgmt/registry.go
Comment (ryantxu) [reply pkg/services/featuremgmt/registry.go:86]: In my original draft, I tried to fix that all up... but decided to punt until future PRs :)

File: pkg/services/featuremgmt/features.go
Comment (ryantxu) [reply pkg/services/featuremgmt/features.go:21]: TBD -- but I think we can/should potentially use feature flags for things that are in development *and* releasable.  If this is all statically defined on startup it does not make much difference, but as we add support for the flag to be set for some users and not others, that seems like something we want to be applied more generally.

I am also nervous that we have many parallel growing lists of features that can be enabled, and hope this more robust structure can give us a powerful single place to improve support

File: packages/grafana-data/src/types/featureToggles.gen.ts
Comment (ryantxu) [reply packages/grafana-data/src/types/featureToggles.gen.ts:15]: Ahh -- i just added a comment that we should uncomment this line before merging.  When it is commented out we get type errors when new frontend keys are created (3 in the last week!)

While developing it is nice to have in strict mode, but for release it should allow any string.

File: pkg/services/featuremgmt/features.go
Comment (ryantxu) [reply pkg/services/featuremgmt/features.go:85]: this is really a placeholder for more complicated work from @bergquist and @oscarkilhed 's hackathon project.  See:
https://github.com/grafana/grafana/blob/experimentals/feature-toggle-by-headers/pkg/experiments/service.go#L141

File: pkg/services/featuremgmt/service.go
Comment (bergquist) [inline pkg/services/featuremgmt/service.go:55]:
```suggestion
	configfile := filepath.Join(cfg.HomePath, "conf", "features.yaml")
```
For this to work we also need to create the file and include it in debian packages etc etc.

**Review by bergquist — COMMENTED** [review]
A few nits. 

I think this is a step it the right direction and after some minor changes we can merge it. But I think we should consider the `features.yaml` feature unstable for a while. I want to have a bit more consensus about how we can operate this on scale first :)

**Review by mckn — COMMENTED** [review]
Frontend code looks good! Great work with this @ryantxu

**Comment by ryantxu** [comment]
@DanCech @bergquist @xlson

I think this is ready for a final review/approval.  The enterprise PR is passing locally, but I think this has to be merged first before CI will help verify :grimacing:

**Comment by DanCech** [comment]
If you have an enterprise branch with the same name CI will run test there against this branch instead of main

File: pkg/services/featuremgmt/manager.go
Comment (DanCech) [inline pkg/services/featuremgmt/manager.go:186]: `flags` doesn't appear to be used

**Review by bergquist — APPROVED** [review]
I think we should continue working on this but I think we can merge this and move forward in another PR. Keeping this PR up to date will just be a massive drag.

**Review by marefr — APPROVED** [review]
LGTM

**Review by DanCech — APPROVED** [review]
Looks good!

**Pull request merged**
